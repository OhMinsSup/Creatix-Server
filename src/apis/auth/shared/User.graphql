type User {
    id: ID!
    username: String!
    email: String!
    created_at: String!
    updated_at: String!
}