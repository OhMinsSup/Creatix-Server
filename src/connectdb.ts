import { getConnectionOptions, createConnection } from 'typeorm';
import dotenv from 'dotenv';
dotenv.config();

export const createConnect = async (dropSchema: boolean = false) => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);
  return createConnection({ ...connectionOptions, name: 'default', dropSchema });
};

export const createConnectProd = async () => {
  const db = await createConnection({
    type: 'postgres',
    url: process.env.DATABASE_URL,
    ssl: true,
    synchronize: true,
    entities: [__dirname + '/entity/**/*.js']
  });
  return db;
};
