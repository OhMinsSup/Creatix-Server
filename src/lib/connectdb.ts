import { getConnectionOptions, createConnection } from 'typeorm';

export const createDevConnection = async () => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);
  return createConnection({ ...connectionOptions, name: 'default' });
};

export const createTestingConnection = async (resetDB: boolean = false) => {
  const connectionOptions = await getConnectionOptions(process.env.NODE_ENV);
  return createConnection({
    ...connectionOptions,
    name: 'default',
    synchronize: resetDB,
    dropSchema: resetDB
  });
};
