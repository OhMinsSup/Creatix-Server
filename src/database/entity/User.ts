import {
  Entity,
  PrimaryGeneratedColumn,
  Index,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  getRepository,
  BeforeInsert,
  BeforeUpdate
} from 'typeorm';
import AuthToken from './AuthToken';
import { generateToken } from '../../lib/token';
import { hash } from '../../lib/utils';

@Entity()
class User {
  @PrimaryGeneratedColumn('uuid')
  public id: string;

  @Index()
  @Column({ unique: true, type: 'varchar' })
  public username: string;

  @Index()
  @Column({ unique: true, type: 'varchar' })
  public email: string;

  @Column('varchar')
  public password: string;

  @Column('timestampz')
  @CreateDateColumn()
  public created_at: Date;

  @Column('timestamptz')
  @UpdateDateColumn()
  public updated_at: Date;

  public comparePassword(password: string): boolean {
    const hashedPassword = hash(password);
    return this.password === hashedPassword;
  }

  @BeforeInsert()
  @BeforeUpdate()
  public savePassword(): void {
    if (this.password) {
      const hashedPassword = hash(this.password);
      this.password = hashedPassword;
    }
  }

  public async generateUserToken() {
    const authToken = new AuthToken();
    authToken.fk_user_id = this.id;
    await getRepository(AuthToken).save(authToken);
    // refresh token is valid for 30days
    const refreshToken = await generateToken(
      {
        user_id: this.id,
        token_id: authToken.id
      },
      {
        subject: 'refresh_token',
        expiresIn: '30d'
      }
    );

    const accessToken = await generateToken(
      {
        user_id: this.id
      },
      {
        subject: 'access_token',
        expiresIn: '1h'
      }
    );

    return {
      refreshToken,
      accessToken
    };
  }

  public async refreshUserToken(
    tokenId: string,
    refreshTokenExp: number,
    originalRefreshToken: string
  ) {
    const now = new Date().getTime();
    const diff = refreshTokenExp * 1000 - now;
    console.log('refreshing..');
    let refreshToken = originalRefreshToken;
    // renew refresh token if remaining life is less than 15d
    if (diff < 1000 * 60 * 60 * 24 * 15) {
      console.log('refreshing refreshToken');
      refreshToken = await generateToken(
        {
          user_id: this.id,
          token_id: tokenId
        },
        {
          subject: 'refresh_token',
          expiresIn: '30d'
        }
      );
    }
    const accessToken = await generateToken(
      {
        user_id: this.id
      },
      {
        subject: 'access_token',
        expiresIn: '1h'
      }
    );
    return { refreshToken, accessToken };
  }
}

export default User;
