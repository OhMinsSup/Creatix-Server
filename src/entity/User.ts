import {
  Entity,
  PrimaryGeneratedColumn,
  Index,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  getRepository
} from 'typeorm';
import AuthToken from './AuthToken';
import { generateToken } from '../lib/token';

@Entity()
class User {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Index()
  @Column({ unique: true, type: 'varchar' })
  username: string;

  @Index()
  @Column({ unique: true, type: 'varchar' })
  email: string;

  @Column('timestampz')
  @CreateDateColumn()
  created_at: Date;

  @Column('timestamptz')
  @UpdateDateColumn()
  updated_at: Date;

  async generateUserToken() {
    const authToken = new AuthToken();
    authToken.fk_user_id = this.id;
    await getRepository(AuthToken).save(authToken);

    // refresh token is valid for 30days
    const refreshToken = await generateToken(
      {
        user_id: this.id,
        token_id: authToken.id
      },
      {
        subject: 'refresh_token',
        expiresIn: '30d'
      }
    );

    const accessToken = await generateToken(
      {
        user_id: this.id
      },
      {
        subject: 'access_token',
        expiresIn: '1h'
      }
    );

    return {
      refreshToken,
      accessToken
    };
  }

  async refreshUserToken(tokenId: string, refreshTokenExp: number, originalRefreshToken: string) {
    const now = new Date().getTime();
    const diff = refreshTokenExp * 1000 - now;
    console.log('refreshing..');
    let refreshToken = originalRefreshToken;

    // renew refresh token if remaining life is less than 15d
    if (diff < 1000 * 60 * 60 * 24 * 15) {
      console.log('refreshing refreshToken');
      refreshToken = await generateToken(
        {
          user_id: this.id,
          token_id: tokenId
        },
        {
          subject: 'refresh_token',
          expiresIn: '30d'
        }
      );
    }
    const accessToken = await generateToken(
      {
        user_id: this.id
      },
      {
        subject: 'access_token',
        expiresIn: '1h'
      }
    );

    return { refreshToken, accessToken };
  }
}

export default User;
