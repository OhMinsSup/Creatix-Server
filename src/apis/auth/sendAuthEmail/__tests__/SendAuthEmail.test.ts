import { Connection, getRepository } from 'typeorm';
import { createConnect } from '../../../../connectdb';
import SendAuthEamilResolver from '../SendAuthEmail.resolvers';
import User from '../../../../entity/User';

describe('SendAuthEmail', () => {
  let connection: Connection;
  const email = 's0ob1kfsb@tmails.net';
  const obj = {};
  const ctx = {};
  const info = {};

  function setEmail(email?: string) {
    const obj = Object.assign(
      {},
      {
        email: email ? email : ''
      }
    );
    return obj;
  }

  beforeAll(async () => {
    connection = await createConnect(true);
  });

  afterAll(async () => {
    connection.close();
  });

  it('Email is not entered or format is wrong', async () => {
    const args = setEmail();
    const result = await SendAuthEamilResolver.Mutation.SendAuthEmail(obj, args, ctx, info);
    expect(result).toEqual({
      ok: false,
      error: '404_EMAIL_NOT_MATCH_FORMAT'
    });
  });

  it('Email member authentication', async () => {
    const args = setEmail(email);

    const userRegisterd = await getRepository(User).findOne({
      where: {
        email: args.email
      }
    });

    const result = await SendAuthEamilResolver.Mutation.SendAuthEmail(obj, args, ctx, info);
    expect(result).toEqual({
      ok: true,
      error: null,
      registered: !!userRegisterd
    });
  });

  it('SendAuthEmail 500 Error Throw', async () => {
    const args = setEmail(email);
    const result = await SendAuthEamilResolver.Mutation.SendAuthEmail(obj, args, ctx, info);
    expect(result).toThrowErrorMatchingSnapshot();
  });
});
