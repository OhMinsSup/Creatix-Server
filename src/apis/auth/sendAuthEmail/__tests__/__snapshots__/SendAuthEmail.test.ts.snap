// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendAuthEmail SendAuthEmail 500 Error Throw 1`] = `"received is not a function"`;
